### ISSUE TITLE
Data Preparation

### ISSUE THREAD

**Issue by Shanzhaguoo** [yvanliang/DriveEditor#12]
Excellent work! Now I want to try switching to a different dataset (e.g., the Waymo dataset) for inference. How can I obtain data in a format similar to data.pkl from the raw data? Could you provide the relevant preprocessing script? Thanks!

**Comment by Shanzhaguoo** [comment]
For the other dataset, I wrote a script for the deletion function once you have the image sequence and masks already:

```
import math
import os
import tempfile
import cv2
import torch
import numpy as np
import gradio as gr
import imageio as imageio_v1
from einops import repeat, rearrange
from omegaconf import OmegaConf
from torchvision import transforms
from torchvision.transforms import InterpolationMode
from PIL import Image
from sgm import instantiate_from_config


class DeletionGUI:
    def __init__(
            self,
            num_frames: int = 10,
            step: int = 25,
            out_size: tuple = (576, 1024),
    ):
        model_config = "configs/sample.yaml"
        self.out_size = out_size
        self.num_frames = num_frames
        self.num_frames_3d = 21
        self.device = "cuda"
        
        # Load model
        self.model = load_model(
            model_config,
            self.device,
            num_steps=step,
            num_frames=num_frames,
            verbose=True,
        )
        
        # Initialize data containers
        self.images = []
        self.masks = []
        self.im_result = []
        
        # Transforms
        self.to_tensor = transforms.ToTensor()
        self.transform_img = transforms.Compose([
            transforms.ToTensor(),
            transforms.Resize(self.out_size, antialias=True),
            transforms.Lambda(lambda x: x * 2.0 - 1.0),
        ])
        self.transform_mask = transforms.Compose([
            transforms.Resize([int(s / 8) for s in out_size], interpolation=InterpolationMode.NEAREST),
            transforms.Lambda(lambda x: x * 2.0 - 1.0),
        ])

    def run(self):
        with gr.Blocks(title="DriveEditor Deletion Tool") as demo:
            gr.HTML("""<h1 style="text-align: center;">DriveEditor: Object Deletion in Driving Scenes</h1>""")
            
            with gr.Row():
                with gr.Column():
                    with gr.Group():
                        gr.Markdown("**Data Input**")
                        image_input = gr.File(
                            label="Upload Image Sequence", 
                            file_count="multiple",
                            file_types=["image"]
                        )
                        mask_input = gr.File(
                            label="Upload Mask Sequence (same order as images)", 
                            file_count="multiple",
                            file_types=["image"]
                        )
                        load_btn = gr.Button(value="Load Data")
                        
                    with gr.Group():
                        gr.Markdown("**Preview**")
                        input_preview = gr.Video(
                            label="Input with Masks", 
                            interactive=False, 
                            loop=True, 
                            autoplay=True,
                            show_download_button=False, 
                            show_share_button=False, 
                            format="mp4"
                        )
                
                with gr.Column():
                    with gr.Group():
                        gr.Markdown("**Deletion Result**")
                        result_viz = gr.Video(
                            label="Result Visualization", 
                            interactive=False, 
                            loop=True, 
                            autoplay=True,
                            show_download_button=False, 
                            show_share_button=False, 
                            format="mp4"
                        )
                        run_btn = gr.Button(value="Generate Deletion")
                    
                    with gr.Group():
                        gr.Markdown("**Options**")
                        decoding_t = gr.Number(
                            1, 
                            label="Images to decode at one time", 
                            minimum=1, 
                            maximum=10, 
                            step=1
                        )
                        seed_input = gr.Number(
                            42, 
                            label="Seed", 
                            minimum=1, 
                            maximum=1000, 
                            step=1
                        )
                        mask_dilation = gr.Slider(
                            minimum=0, 
                            maximum=20, 
                            value=5,
                            step=1,
                            label="Mask dilation (pixels)"
                        )
                    
                    gr.Markdown("""
                    **Instructions:**
                    1. Upload your image sequence (e.g., frame001.jpg, frame002.jpg, ...)
                    2. Upload corresponding mask sequence in the same order
                    3. Click "Load Data" to preview
                    4. Adjust parameters if needed
                    5. Click "Generate Deletion" to remove objects
                    
                    **Note:** Masks should be binary images where white (255) indicates the object to delete.
                    """)
            
            @load_btn.click(inputs=[image_input, mask_input, mask_dilation], 
                           outputs=[input_preview])
            def load_data(image_files, mask_files, dilation):
                if image_files is None or mask_files is None:
                    gr.Warning("Please upload both images and masks")
                    return None
                
                if len(image_files) != len(mask_files):
                    gr.Warning(f"Number of images ({len(image_files)}) must match number of masks ({len(mask_files)})")
                    return None
                
                if len(image_files) > self.num_frames:
                    gr.Warning(f"Maximum {self.num_frames} frames supported. Using first {self.num_frames} frames.")
                    image_files = image_files[:self.num_frames]
                    mask_files = mask_files[:self.num_frames]
                
                # Load and process images and masks
                self.images = []
                self.masks = []
                preview_frames = []
                
                for img_file, mask_file in zip(image_files, mask_files):
                    # Load image
                    img = Image.open(img_file.name).convert('RGB')
                    img = img.resize((self.out_size[1], self.out_size[0]), Image.LANCZOS)
                    img_array = np.array(img)
                    self.images.append(img_array)
                    
                    # Load mask
                    mask = Image.open(mask_file.name).convert('L')
                    mask = mask.resize((self.out_size[1], self.out_size[0]), Image.NEAREST)
                    mask_array = np.array(mask)
                    
                    # Apply dilation if specified
                    if dilation > 0:
                        kernel = np.ones((dilation*2+1, dilation*2+1), np.uint8)
                        mask_array = cv2.dilate(mask_array, kernel, iterations=1)
                    
                    # Threshold to binary
                    mask_array = (mask_array > 127).astype(np.float32)
                    self.masks.append(mask_array)
                    
                    # Create preview with mask overlay
                    preview = img_array.copy()
                    mask_colored = np.zeros_like(preview)
                    mask_colored[:, :, 0] = mask_array * 255  # Red channel
                    preview = cv2.addWeighted(preview, 0.7, mask_colored, 0.3, 0)
                    preview_frames.append(preview)
                
                # Pad with zeros if less than num_frames
                while len(self.images) < self.num_frames:
                    self.images.append(np.zeros_like(self.images[-1]))
                    self.masks.append(np.zeros_like(self.masks[-1]))
                
                # Create preview video
                preview_video = create_video(preview_frames)
                return preview_video
            
            @run_btn.click(inputs=[decoding_t, seed_input], 
                          outputs=[result_viz])
            def generate_deletion(decoding_t, seed_value):
                if len(self.images) == 0:
                    gr.Warning("Please load data first")
                    return None
                
                set_seed(int(seed_value))
                result = self.predict_deletion(decoding_t)
                return result
        
        demo.launch(share=True)
    
    def predict_deletion(self, decoding_t):
        """Generate deletion result"""
        # Prepare inputs for deletion
        cond_im_all = []
        mask_all = []
        
        for img, mask in zip(self.images[:self.num_frames], self.masks[:self.num_frames]):
            # Process image
            cond_im = self.to_tensor(img) * 2. - 1.
            
            # Apply mask to image (set masked regions to zero)
            mask_expanded = np.expand_dims(mask, axis=0)
            mask_expanded = np.repeat(mask_expanded, 3, axis=0)
            cond_im = cond_im * (1 - torch.from_numpy(mask_expanded).float())
            
            cond_im_all.append(cond_im)
            mask_all.append(torch.tensor(mask).unsqueeze(0))
        
        # Stack and transform
        cond_im_all = torch.stack(cond_im_all, dim=0)
        mask_all = self.transform_mask(torch.stack(mask_all, dim=0))
        
        # Create dummy inputs for unused features
        position_all = [{} for _ in range(self.num_frames)]
        obj_im_clip = torch.ones(3, 224, 224)
        cond_im_3d = torch.ones(1, 3, 576, 576)
        box_im_all = torch.ones(self.num_frames, 6, *self.out_size) * -1.
        mask_fuse_all = torch.zeros_like(mask_all)
        elevations_all = torch.zeros(self.num_frames_3d).unsqueeze(-1)
        azimuths_all = torch.zeros(self.num_frames_3d).unsqueeze(-1)
        indices_3d_all = torch.zeros(self.num_frames_3d).unsqueeze(-1)
        valid_mask = torch.zeros(self.num_frames)
        obj_ratio = torch.zeros(self.num_frames)
        
        # Prepare shape
        H, W = cond_im_all.shape[2:]
        shape = (self.num_frames, 4, H // 8, W // 8)
        shape_3d = (self.num_frames_3d, 4, 576 // 8, 576 // 8)
        
        # Create value dict
        value_dict = dict(
            motion_bucket_id=127,
            fps_id=10,
            cond_aug=torch.tensor(0.02).repeat(self.num_frames),
            cond_aug_3d=torch.tensor(1e-5).repeat(self.num_frames_3d),
            polars_rad_3d=elevations_all,
            azimuths_rad_3d=azimuths_all,
            indices_3d=indices_3d_all,
            obj_pos=position_all,
            cond_frames_without_noise=[cond_im_all[0:1], obj_im_clip.unsqueeze(0)],
            cond_frames=cond_im_all + torch.tensor(0.02) * torch.randn_like(cond_im_all),
            cond_frames_without_noise_3d=cond_im_3d,
            cond_frames_3d=cond_im_3d + torch.tensor(1e-5) * torch.randn_like(cond_im_3d),
            mask_concat=mask_all,
            mask_fuse=mask_fuse_all,
            depth=box_im_all,
        )
        
        # Run model inference
        with torch.no_grad():
            with torch.autocast(device_type="cuda", dtype=torch.float16):
                batch, batch_uc = get_batch(
                    get_unique_embedder_keys_from_conditioner(self.model.conditioner) +
                    get_unique_embedder_keys_from_conditioner(self.model.conditioner_3d),
                    value_dict,
                    [1, self.num_frames],
                    device=self.device,
                )
                
                c, uc = self.model.conditioner.get_unconditional_conditioning(
                    batch,
                    batch_uc=batch_uc,
                    force_uc_zero_embeddings=[
                        "cond_frames",
                        "cond_frames_without_noise",
                    ],
                )
                
                c_3d, uc_3d = self.model.conditioner_3d.get_unconditional_conditioning(
                    batch,
                    batch_uc=batch_uc,
                    force_uc_zero_embeddings=[
                        "cond_frames_3d",
                        "cond_frames_without_noise_3d",
                    ],
                )
                
                # Process conditioning
                for k in ["crossattn"]:
                    uc[k] = repeat(uc[k], "b ... -> b t ...", t=self.num_frames)
                    uc[k] = rearrange(uc[k], "b t ... -> (b t) ...", t=self.num_frames)
                    c[k] = repeat(c[k], "b ... -> b t ...", t=self.num_frames)
                    c[k] = rearrange(c[k], "b t ... -> (b t) ...", t=self.num_frames)
                
                for k in ["crossattn", "concat"]:
                    uc_3d[k] = repeat(uc_3d[k], "b ... -> b t ...", t=self.num_frames_3d)
                    uc_3d[k] = rearrange(uc_3d[k], "b t ... -> (b t) ...", t=self.num_frames_3d)
                    c_3d[k] = repeat(c_3d[k], "b ... -> b t ...", t=self.num_frames_3d)
                    c_3d[k] = rearrange(c_3d[k], "b t ... -> (b t) ...", t=self.num_frames_3d)
                
                # Move position data to device
                for d in position_all:
                    for k, v in d.items():
                        if isinstance(v, torch.Tensor):
                            d[k] = v.to(self.device)
                
                # Create random noise
                randn = torch.randn(shape, device=self.device)
                randn_3d = torch.randn(shape_3d, device=self.device)
                
                # Additional model inputs
                additional_model_inputs = {}
                additional_model_inputs["image_only_indicator"] = torch.zeros(2, self.num_frames).to(self.device)
                additional_model_inputs["image_only_indicator_3d"] = torch.zeros(2, self.num_frames_3d).to(self.device)
                additional_model_inputs["num_video_frames"] = self.num_frames
                additional_model_inputs["num_video_frames_3d"] = self.num_frames_3d
                additional_model_inputs["valid_mask"] = torch.concat([valid_mask] * 2).to(self.device)
                additional_model_inputs["mask_fuse"] = torch.concat([mask_fuse_all] * 2).to(self.device)
                additional_model_inputs["obj_ratio"] = torch.concat([obj_ratio] * 2).to(self.device)
                additional_model_inputs['obj_pos'] = position_all * 2
                additional_model_inputs['indices_3d'] = torch.concat(
                    [
                        indices_3d_all,
                        indices_3d_all + self.num_frames_3d
                    ]
                ).to(self.device)
                
                # Move conditioning to device
                for d in (c, uc, c_3d, uc_3d):
                    for k, v in d.items():
                        if isinstance(v, torch.Tensor):
                            d[k] = v.to(self.device)
                        elif isinstance(v, list):
                            d[k] = [vv.to(self.device) for vv in v]
                
                # Define denoiser
                def denoiser(input, sigma, c, input_3d, sigma_3d, c_3d):
                    return self.model.denoiser(
                        self.model.model, input, input_3d, sigma, sigma_3d, c, c_3d, **additional_model_inputs
                    )
                
                # Sample
                samples_z, _ = self.model.sampler(denoiser, randn, randn_3d,
                                                cond=c, uc=uc, cond_3d=c_3d, uc_3d=uc_3d)
                
                self.model.en_and_decode_n_samples_a_time = decoding_t
                
                # Decode
                torch.cuda.empty_cache()
                samples_x = self.model.decode_first_stage(samples_z)
                samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
                vid = (
                    (rearrange(samples, "t c h w -> t h w c") * 255)
                    .cpu()
                    .numpy()
                    .astype(np.uint8)
                )
                
                self.im_result = [np.ascontiguousarray(v) for v in vid]
                return create_video(self.im_result)


def create_video(images):
    """Create a video from a list of images"""
    video_filename = os.path.join(tempfile.gettempdir(), 'output.mp4')
    
    output_writer = imageio_v1.get_writer(
        video_filename,
        fps=7,
        codec="libx264",
        macro_block_size=None,
        ffmpeg_params=[
            "-crf", "20",
            "-preset", "slow",
            "-pix_fmt", "yuv420p"
        ],
    )
    for img in images:
        output_writer.append_data(img)
    output_writer.close()
    
    return video_filename


def load_model(
        config: str,
        device: str,
        num_steps: int,
        num_frames: int,
        verbose: bool = False,
):
    """Load the model"""
    config = OmegaConf.load(config)
    config.model.params.sampler_config.params.verbose = verbose
    config.model.params.sampler_config.params.num_steps = num_steps
    config.model.params.sampler_config.params.guider_config.params.num_frames = num_frames
    model = instantiate_from_config(config.model).to(device).eval()
    return model


def get_unique_embedder_keys_from_conditioner(conditioner):
    """Get unique embedder keys from conditioner"""
    return list(set([x.input_key for x in conditioner.embedders]))


def get_batch(keys, value_dict, N, device):
    """Prepare batch for model"""
    batch = {}
    batch_uc = {}
    
    for key in keys:
        if key == "fps_id":
            batch[key] = (
                torch.tensor([value_dict["fps_id"]])
                .to(device)
                .repeat(int(math.prod(N)))
            )
        elif key == "motion_bucket_id":
            batch[key] = (
                torch.tensor([value_dict["motion_bucket_id"]])
                .to(device)
                .repeat(int(math.prod(N)))
            )
        elif key == "cond_frames" or key == "cond_frames_3d":
            batch[key] = value_dict[key].to(device)
        elif key == "cond_frames_without_noise":
            batch[key] = [v.to(device) for v in value_dict["cond_frames_without_noise"]]
        else:
            batch[key] = value_dict[key].to(device)
    
    for key in batch.keys():
        if key not in batch_uc and isinstance(batch[key], torch.Tensor):
            batch_uc[key] = torch.clone(batch[key])
        elif key not in batch_uc and isinstance(batch[key], list):
            batch_uc[key] = [torch.clone(v) for v in batch[key]]
    
    return batch, batch_uc


def set_seed(seed):
    """Set random seed for reproducibility"""
    torch.manual_seed(seed)
    np.random.seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False


if __name__ == "__main__":
    gui = DeletionGUI()
    gui.run()
```
